doc/protocol.md

```diff
@@ -674,6 +674,193 @@ sent by the server might have the following additional fields:
+For instanced metrics, the array contains another array with samples
+for each instance, in the same order as reported in the "instances"
+field of the most recent 'meta' message.
+
+In order to gain efficiency, 'data' messages are usually compressed.
+This is done by only transmitting the differences from one point in
+time to the next.
+
+If a value for a metric or a instance is the same as at the previous
+point in time, the channel transmits a "null" value instead.
+Additionally, "null" values at the end of an array are suppressed by
+transmitting a shorter array.
+
+For example, say the samples for three points in time are
+
+  1: [ 21354, [ 5,  5, 5 ], 100 ]
+  2: [ 21354, [ 5, 15, 5 ], 100 ]
+  3: [ 21354, [ 5, 15, 5 ], 100 ]
+
+then the channel will send these array instead:
+
+  1: [ 21354, [    5,  5, 5 ], 100 ]
+  2: [  null, [ null, 15 ] ]
+  3: [  null, [ ] ]
+
+# PCP metric sources
+
+You specify the desired metrics as an array of objects, where each
+object describes one metric.  For example:
+
+  [ { name: "kernel.all.cpu.user",
+      type: "number",
+      units: "millisec",
+      semantics: "counter"
+    },
+    ...
+  ]
+
+A metric description is used to describe the expected behavior of the
+metric.  For example, you can specify that you expect numbers for a
+given metric and can then be sure that the source will not
+unexpectedly deliver strings to you.
+
+A metric description can have the following fields:
+
+ * "name" (string): The name of the metric.  Use "pminfo -L" to get a
+   list of available PCP metric names for a "direct" source, for
+   example.  If no metric of this name exists, the channel is closed
+   without delivering any message.
+
+ * "type" (string, optional): The expected type of the metric, either
+   "number" or "string".  If no type is specified, any type is allowed
+   (unless "units" is also specified, see below).  If the metric is
+   not of the given type, the channel is closed.
+
+ * "semantics" (string, optional): The expected semantics of the
+   metric, one of "counter", "instant", or "discrete".  If the metric
+   does not have the given semantics, the channel is closed.
+
+ * "units" (string, optional): The units that values for this metric
+   should be delivered in.  If the given metric can not be converted
+   into these units, the channel is closed.  The format of the string
+   is the same as the one used by "pminfo -d".
+
+   When specifying "units", "type" defaults to "number".
+
+The 'meta' messages for PCP sources also contain these fields:
+
+ * "types" (array of strings): The type of the values reported for
+   each metric, either "number" or "string".
+
+ * "semantics" (array of strings): The semantics of each metric, one
+   of "counter", "instant", or "discrete".  Counter metrics are
+   treated specially by a metrics channel, see below.
+
+ * "units" (array of strings): The units for the values reported for
+   each metric.
```

Comment:
Shouldn't these be similar objects with keys as the metrics specified when opening the channel? Why are these arrays of primitives, and the open message contains arrays of objects?